File: exampleadapters/src/main/java/foo/bar/example/asafadapters/feature/playlist/PlaylistSimpleModel.java
package foo.bar.example.asafadapters.feature.playlist;


import java.util.ArrayList;
import java.util.List;

import co.early.asaf.core.Affirm;
import co.early.asaf.core.WorkMode;
import co.early.asaf.core.logging.Logger;
import co.early.asaf.core.observer.ObservableImp;

import static foo.bar.example.asafadapters.feature.playlist.RandomTrackGeneratorUtil.generateRandomColourResource;

/**
 *
 */
public class PlaylistSimpleModel extends ObservableImp{

    private static String TAG = PlaylistSimpleModel.class.getSimpleName();

    private final Logger logger;
    private List<Track> trackList = new ArrayList<>();


    public PlaylistSimpleModel(WorkMode workMode, Logger logger) {
        super(workMode, logger);
        this.logger = Affirm.notNull(logger);
    }

    public void addNewTrack(){
        logger.i(TAG, "addNewTrack()");
        trackList.add(new Track(generateRandomColourResource()));
        notifyObservers();
    }

    public void removeTrack(int index){
        logger.i(TAG, "removeTrack() "+index);
        checkIndex(index);
        trackList.remove(index);
        notifyObservers();
    }

    public void removeAllTracks() {
        logger.i(TAG, "removeAllTracks()");
        trackList.clear();
        notifyObservers();
    }

    public void increasePlaysForTrack(int index){
        logger.i(TAG, "increasePlaysForTrack() "+index);
        getTrack(index).increasePlaysRequested();
        notifyObservers();
    }

    public void decreasePlaysForTrack(int index){
        logger.i(TAG, "decreasePlaysForTrack() "+index);
        getTrack(index).decreasePlaysRequested();
        notifyObservers();
    }

    public Track getTrack(int index){
        checkIndex(index);
        return trackList.get(index);
    }

    public int getTrackListSize(){
        return trackList.size();
    }

    public void add5NewTracks() {
        logger.i(TAG, "add5NewTracks()");
        List<Track> newTracks = new ArrayList<>();
        for (int ii=0; ii<5; ii++){
            newTracks.add(new Track(generateRandomColourResource()));
        }
        trackList.addAll(0, newTracks);
        notifyObservers();
    }

    public void remove5Tracks() {
        logger.i(TAG, "remove5Tracks()");
        if (getTrackListSize()>4){
            trackList.subList(0, 5).clear();
            notifyObservers();
        }
    }

    private void checkIndex(int index) {
        if (trackList.size() == 0) {
            throw new IndexOutOfBoundsException("tracklist has no items in it, can not get index:" + index);
        } else if (index < 0 || index > trackList.size() - 1) {
            throw new IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (trackList.size() - 1) + " not:" + index);
        }
    }

}

File: example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/feature/playlist/PlaylistSimpleModel.kt
package foo.bar.example.foreadapterskt.feature.playlist


import co.early.fore.core.WorkMode
import co.early.fore.core.observer.Observable
import co.early.fore.kt.core.logging.Logger
import co.early.fore.kt.core.observer.ObservableImp
import foo.bar.example.foreadapterskt.feature.playlist.RandomTrackGeneratorUtil.generateRandomColourResource
import java.util.ArrayList

/**
 * Copyright © 2019 early.co. All rights reserved.
 */
class PlaylistSimpleModel(
        workMode: WorkMode,
        private val logger: Logger
) : Observable by ObservableImp(workMode, logger) {

    private val trackList = ArrayList<Track>()

    val trackListSize: Int
        get() = trackList.size


    fun addNewTrack() {
        logger.i("addNewTrack()")
        trackList.add(Track(generateRandomColourResource()))
        notifyObservers()
    }

    fun removeTrack(index: Int) {
        logger.i("removeTrack() $index")
        checkIndex(index)
        trackList.removeAt(index)
        notifyObservers()
    }

    fun removeAllTracks() {
        logger.i(LOG_TAG, "removeAllTracks()")
        trackList.clear()
        notifyObservers()
    }

    fun increasePlaysForTrack(index: Int) {
        logger.i(LOG_TAG, "increasePlaysForTrack() $index")
        getTrack(index).increasePlaysRequested()
        notifyObservers()
    }

    fun decreasePlaysForTrack(index: Int) {
        logger.i(LOG_TAG, "decreasePlaysForTrack() $index")
        getTrack(index).decreasePlaysRequested()
        notifyObservers()
    }

    fun getTrack(index: Int): Track {
        checkIndex(index)
        return trackList[index]
    }

    fun add5NewTracks() {
        logger.i(LOG_TAG, "add5NewTracks()")
        val newTracks = ArrayList<Track>()
        for (ii in 0..4) {
            newTracks.add(Track(generateRandomColourResource()))
        }
        trackList.addAll(0, newTracks)
        notifyObservers()
    }

    fun remove5Tracks() {
        logger.i(LOG_TAG, "remove5Tracks()")
        if (trackListSize > 4) {
            trackList.subList(0, 5).clear()
            notifyObservers()
        }
    }

    private fun checkIndex(index: Int) {
        if (trackList.size == 0) {
            throw IndexOutOfBoundsException("tracklist has no items in it, can not get index:$index")
        } else if (index < 0 || index > trackList.size - 1) {
            throw IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (trackList.size - 1) + " not:" + index)
        }
    }

    companion object {
        private val LOG_TAG = PlaylistSimpleModel::class.java.simpleName
    }

}
